subprojects/core/src/integTest/groovy/org/gradle/api/provider/PropertyStateIntegrationTest.groovy

```diff
@@ -0,0 +1,372 @@
+            import org.gradle.api.tasks.OutputFiles
+
+            class MyTask extends DefaultTask {
+                private PropertyState<Boolean> enabled = project.property(Boolean)
+                private PropertyState<FileCollection> outputFiles = project.property(FileCollection)
+                
+                MyTask() {
+                    enabled.set(false)
+                    outputFiles.set(project.files())
+                }
+
+                @Input
+                boolean getEnabled() {
+                    enabled.get()
+                }
+                
+                void setEnabled(PropertyState<Boolean> enabled) {
+                    this.enabled = enabled
+                }
+                
+                void setEnabled(boolean enabled) {
+                    this.enabled.set(enabled)
+                }
+                
+                @OutputFiles
+                FileCollection getOutputFiles() {
+                    outputFiles.get()
+                }
+
+                void setOutputFiles(PropertyState<FileCollection> outputFiles) {
+                    this.outputFiles = outputFiles
+                }
+                
+                void setOutputFiles(FileCollection outputFiles) {
+                    this.outputFiles.set(outputFiles)
+                }
+
+                @TaskAction
+                void resolveValue() {
+                    if (getEnabled()) {
+                        getOutputFiles().each {
+                            it.text = '$OUTPUT_FILE_CONTENT'
+                        }
+                    }
+                }
+            }
+        """
+    }
+
+    static String customJavaBasedTaskType() {
+        """
+            import org.gradle.api.DefaultTask;
+            import org.gradle.api.file.ConfigurableFileCollection;
+            import org.gradle.api.provider.PropertyState;
+            import org.gradle.api.tasks.TaskAction;
+            import org.gradle.api.tasks.Input;
+            import org.gradle.api.tasks.OutputFiles;
+
+            import java.io.BufferedWriter;
+            import java.io.File;
+            import java.io.FileWriter;
+            import java.io.IOException;
+            import java.util.concurrent.Callable;
+        
+            public class MyTask extends DefaultTask {
+                private PropertyState<Boolean> enabled;
+                private PropertyState<ConfigurableFileCollection> outputFiles;
+
+                public MyTask() {
+                    enabled = getProject().property(Boolean.class);
+                    enabled.set(false);
+                    outputFiles = getProject().property(ConfigurableFileCollection.class);
+                    outputFiles.set(getProject().files());
+                }
+
+                @Input
+                public boolean getEnabled() {
+                    return enabled.get();
+                }
+                
+                public void setEnabled(PropertyState<Boolean> enabled) {
```

Comment:
should be `setEnabled(Provider<Boolean> enabled) { this.enabled.set(enabled); }`